src/protocol/parser.rs

```diff
@@ -0,0 +1,130 @@
+            self.parse_error(message)?;
+        }
+        Ok(())
+    }
+
+    fn peek_chars(&mut self, expected: Vec<Option<char>>, message: &str) -> R<()> {
+        let peeked = self.input.get(0);
+        if !expected.contains(&peeked.cloned()) {
+            self.parse_error(message)?;
+        }
+        Ok(())
+    }
+
+    fn parse_char(&mut self, excluded: &[char]) -> R<Option<char>> {
+        Ok(match self.input.pop_front() {
+            None => None,
+            Some('\\') => Some(match self.input.pop_front() {
+                None => self.parse_error("a backslash must be followed by a character")?,
+                Some('"') => '"',
+                Some('n') => '\n',
+                Some(' ') => ' ',
+                Some('\\') => '\\',
+                Some(char) => {
+                    self.parse_error(&format!("unknown escaped character {}", char))?
+                }
+            }),
+            Some(char) => {
+                if excluded.contains(&char) {
+                    self.input.push_front(char);
+                    None
+                } else {
+                    Some(char)
+                }
+            }
+        })
+    }
+
+    fn collect_chars_until(&mut self, excluded: &[char]) -> R<String> {
+        let mut result = "".to_string();
+        while let Some(char) = self.parse_char(excluded)? {
+            result.push(char);
+        }
+        Ok(result)
+    }
+
+    fn parse_word(&mut self) -> R<Option<String>> {
+        self.skip_spaces();
+        Ok(match self.input.get(0) {
+            None => None,
+            Some('"') => {
+                self.skip_char('"', "shouldn't happen")?;
+                let word = self.collect_chars_until(&['"'])?;
+                self.skip_char('"', "unmatched quotes")?;
+                self.peek_chars(
+                    vec![Some(' '), None],
+                    "closing quotes must be followed by a space",
+                )?;
+                Some(word)
+            }
+            Some(_) => {
+                let result = self.collect_chars_until(&[' ', '"'])?;
+                self.peek_chars(
+                    vec![Some(' '), None],
+                    "opening quotes must be preceeded by a space",
+                )?;
+                Some(result)
+            }
+        })
+    }
+
+    pub fn parse_command(&mut self) -> R<Command> {
+        let executable = match self.parse_word()? {
+            None => self.parse_error("expected: space-separated command and arguments")?,
+            Some(executable) => executable.into_bytes(),
+        };
+        let mut arguments = vec![];
+        loop {
+            match self.parse_word()? {
+                None => break,
+                Some(word) => arguments.push(word.into_bytes()),
+            }
+        }
+        Ok(Command {
+            executable,
+            arguments,
+        })
+    }
+
+    pub fn parse_script_arguments(&mut self) -> R<Vec<String>> {
```

Comment:
here is the new method for parsing script arguments, following the same style as above in `parse_command`